### PR TITLE
Recurrent Neural Network for Text Classification - Advanced Example

### PR DESCRIPTION
**Pull request by DanyEle**
Added Jupyter Notebook code for a Recurrent Neural Network based on two models being learned in a federated way. The RNN is used for text classification as belonging to a certain category. 

The code was also tested with Raspberry Pis using remote worker websockets. 

Federated implementation of: https://pytorch.org/tutorials/intermediate/char_rnn_classification_tutorial.html

### CODE REVIEW

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb [hunk 1]
```diff
@@ -0,0 +1,985 @@
+{
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>&gt; Following distributed training, the resulting model is going to be able to perform operations like: ?</p><p><br></p><p>Missing some text here.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Recommend: to pip install -r PySyft/requirements.txt</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>can you add a space before and after all the "=" at variables declaration?</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Make the first comment in this cell a text cell.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Make the first long comment in this cell a text cell.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Make the first comment in this cell a text cell.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Make the first comment in this cell a text cell.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Make the first comment in this cell a text cell.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Make the first comment in this cell a text cell.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Make the first comment in this cell a text cell.</p>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/2128/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

**Comment by mari-linhares** [comment]
Oh @DanyEle, can you also add at the end of the tutorial the default message about pysyft, like it was done [here](https://render.githubusercontent.com/view/ipynb?commit=8083f73cde594de6de2c0435555dcd160c25be70&enc_url=68747470733a2f2f7261772e67697468756275736572636f6e74656e742e636f6d2f4f70656e4d696e65642f5079537966742f383038336637336364653539346465366465326330343335353535646364313630633235626537302f6578616d706c65732f7475746f7269616c732f50617274253230312532302d2532305468652532304261736963253230546f6f6c732532306f6625323050726976617465253230446565702532304c6561726e696e672e6970796e62&nwo=OpenMined%2FPySyft&path=examples%2Ftutorials%2FPart+1+-+The+Basic+Tools+of+Private+Deep+Learning.ipynb&repository_id=97641933&repository_type=Repository#Congratulations!!!---Time-to-Join-the-Community!).

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb (hunk 1 above)
Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>It was actually referring to the next cell's predictions. I added some more introductory text and a brief explanation about RNNs' training BTW.</p>

Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>To have an actual executable command, I added:</p><p><br></p><p><code>pip install -r "../../../requirements.txt"</code></p>

Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Done!</p>

Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Done!</p>

Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Done, and re-rephrased the explanation about word embeddings.</p>

Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: <p>Done!</p>

**Comment by DanyEle** [comment]
Added the default message about PySyft at the end of the Jupyter Notebook.

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb [hunk 2]
```diff
@@ -0,0 +1,985 @@
...
+    "\n",
+    "    #extract a random element from the list and perform training on it\n",
+    "    for iter in range(1, n_iters + 1):        \n",
+    "        #random_index = random.randint(0, len(list_federated_train_loader) - 1)\n",
+    "        random_index = randomTrainingIndex()\n",
+    "        line_single, category_single = list_federated_train_loader[random_index]\n",
+    "        #print(category_single.copy().get())\n",
+    "        line_name = names_list[random_index]\n",
+    "        model_pointers, loss, output = fw_bw_pass_model(model_pointers, line_single, category_single)\n",
+    "        #model_pointers = fed_avg_every_n_iters(model_pointers, iter, args.federate_after_n_batches)\n",
```
Comment (midokura-silvia) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: So is there already a federated averaging step included? It is commented out here and I didn't see the implementation of the function fed_avg_every_n_iters(). 

For information, there is code to average models already available: see syft.frameworks.torch.federated.utils.federated_avg()

Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: Actually, the function fed_avg_every_n_iters performing federated averaging is there in the last commit. It is indeed based on the `utils.federated_avg()` function,  however I commented the federated averaging part because the model loss increases in correspondence with the averaging function call, causing the overall model loss to diverge over time. By discussing this issue with a professor of mine and with Andrew Trask, the reason for that probably the small batch size = 1 being used in the present example.

**Comment by mari-linhares** [comment]
LGTM, I'll have a closer look later today.

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb [hunk 3]
```diff
@@ -0,0 +1,1133 @@
...
+    "\n",
+    "We will train two Recurrent Neural Networks residing on two remote workers based on a dataset containing approximately 20.000 surnames from 18 languages of origin, and predict to which language a name belongs based on the name's spelling. \n",
+    "\n",
+    "A **character-level RNN** treats words as a series of characters - outputting a prediction and “hidden state” per character, feeding its previous hidden state into each next step. We take the final prediction to be the output, i.e. which class the word belongs to. Hence the training process proceeds sequentially character-by-character\n",
+    "\n",
+    "Following distributed training, the resulting models are going to be able to perform operations like the following ones:"
+   ]
+  },
+  {
+   "cell_type": "raw",
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: I don't know why/how, but there is text leaving the first cell:

image

I think is because this cell has type "raw" instead of "markdown", can you change it to markdown?

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb [hunk 4]
```diff
@@ -0,0 +1,1133 @@
...
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We now need to turn every single **character** in each input line string into a vector, with a \"1\" marking the character present in that very character.\n",
+    "\n",
+    "#Example: character \"a\" = array([[1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: Remove # before example. In markdown this # -> title, which is probably unintended in this case.

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb [hunk 5]
```diff
@@ -0,0 +1,1133 @@
...
+    "            langDataset\n",
+    "            .federate(workers_virtual),\n",
+    "            batch_size=args.batch_size)  \n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 4 - Training!\n",
```
Comment (mari-linhares) [inline examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: Can you use a single pattern for section titles? You started using <number> . <title>, in the end you're using <number> - <title>

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb (hunk 3 above)
Comment (robert-wagner) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: I have this same issue

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb (hunk 5 above)
Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: Done, made it all more coherent

File: examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb (hunk 4 above)
Comment (DanyEle) [reply examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb]: Fixed
